test/distributed/_composable/fsdp/test_fully_shard_training.py

```diff
@@ -423,5 +523,81 @@ def _test_train_shared_params(
             self.assertEqual(losses[0], losses[1])
 
 
+class TestFullyShard2DTraining(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(4, torch.cuda.device_count())
+
+    @skip_if_lt_x_gpu(2)
+    def test_train_parity_2d_mlp(self):
+        # Prefer to test with >=4 GPUs, but for 2 GPUs, use 2-way TP
+        dp_size = 2 if self.world_size > 2 else 1
+        global_mesh = init_device_mesh(
+            "cuda", (dp_size, self.world_size // dp_size), mesh_dim_names=("dp", "tp")
+        )
+        self.run_subtests(
+            {
+                "reshard_after_forward": [False, True],
+                "use_activation_checkpointing": [False, True],
+                "mlp_dim": [3, 16, 17],
+            },
+            functools.partial(self._test_train_parity_2d_mlp, global_mesh),
+        )
```

Comment:
The main reason I have been subtesting by defining a separate `_test...` method is so that we do not have to indent the entire test code due to:
```
for arg1, arg2, ... in itertools.product(...):
    # Actual test code
```
The `run_subtests` helper that hides that `itertools.product` (or similar) takes in a method, which is why there is a separate `_test_train_parity_2d_mlp` method 😢 .

If you have a better solution for how to do subtesting elegantly, I would be happy to migrate the FSDP2 test suite.

Reply:
Also, I have the `global_mesh = init_device_mesh()` only once to reuse across all subtests to avoid the expensive `init_process_group()` calls upon each subtest. Once we re-land the process group caching, this would be able to be simplified (pushing the `global_mesh = init_device_mesh()` call into `_test_train_parity_2d_mlp`).